### PR TITLE
Let plugins decide if events should be processed or not

### LINKED ISSUE #5594: Endless background jobs
### Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find a similar issue

### Describe the bug*

I don't know how to describe it, I'll try my best, please ask if you'r missing something.

### Background
I'm onboarding a small business to InvenTree. It has a 10+ years history in eCommerce (but very small) and moved over to handmade products, so InvenTree will come into play to deal with stock. I decided to import "the past" (re-create stock movements so to say) for some resons. It's partially automated, so it's a fun deep dive into InvenTree.

### Setup
I host it on DS918+ (INTEL Celeron J3455, 1.5 GHz) with 8 GiB RAM. It runs on Docker via Protainer.

### Dimension
Currently I have 10 Parts, 10 Purchase Orders, 137 Sale Orders.  
I would say that's not much.

### Problem
I can create new customers, I can create Sale Orders. I can issue Sale Orders. All this runns quiet smooth. BUT: When I ship the order items, things start to go south.

Per shipment, roughly 200 or 300 background jobs are spawned (while being processed), which take probably 5 minutes to go back down to 0. For sure it decreases every now and then, but also it increases again.

If I ship 3 or 4 orders, the queue is averaging about 700 - 900 queued jobs, that lasts for 20 minutes or so.

### Steps to Reproduce

1. Ship a ton of orders

### Expected behaviour

Maybe it's to blue-eyed, but I would say shipping some orders will just do some database updates. That should be processed within no time.

### Deployment Method

- [X] Docker
- [ ] Bare metal

### Version Information

# Version Information:
InvenTree-Version: 0.12.7
Django Version: 3.2.19
Commit Hash: 6563b4c
Commit Date: 2023-09-16

Database: postgresql
Debug-Mode: False
Deployed using Docker: True
Platform: Linux-4.4.180+-x86_64-with-glibc2.36
Installer: DOC

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.0.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'AutoFinishSalesOrderPlugin', 'slug': 'mj-auto-finish-sales-order', 'version': '0.0.0'}, {'name': 'AutoProvisionSalesOrderPlugin', 'slug': 'mj-auto-provision-sales-order', 'version': '0.0.0'}, {'name': 'OrderImportPlugin', 'slug': 'mj-order-import-plugin', 'version': '0.0.0'}, {'name': 'OrderProfitPlugin', 'slug': 'mj-order-profit-plugin', 'version': '0.0.0'}, {'name': 'OrderWeightPlugin', 'slug': 'mj-order-weight-plugin', 'version': '0.0.0'}]

### Please verify if you can reproduce this bug on the demo site.

- [ ] I can reproduce this bug on the demo site.

### Relevant log output

```shell
08:35:44 [Q] INFO Processed [foxtrot-fourteen-illinois-massachusetts]
08:35:45 [Q] INFO Processed [diet-mango-mirror-triple]
08:35:45 [Q] INFO Processed [fruit-kilo-sixteen-blossom]
08:35:45 [Q] INFO Processed [triple-washington-neptune-fruit]
08:35:45 [Q] INFO Processed [delaware-romeo-early-sierra]
08:35:45 [Q] INFO Processed [foxtrot-football-oven-stream]
08:35:45 [Q] INFO Processed [salami-football-failed-angel]
08:35:45 [Q] INFO Processed [glucose-louisiana-massachusetts-oxygen]
08:35:46 [Q] INFO Processed [indigo-october-two-mockingbird]
08:35:46 [Q] INFO Processed [kentucky-arizona-paris-seven]
08:35:46 [Q] INFO Processed [lactose-lamp-diet-whiskey]
08:35:46 [Q] INFO Processed [october-aspen-bacon-thirteen]
08:35:46 [Q] INFO Processed [lion-failed-oven-nine]
08:35:46 [Q] INFO Processed [quebec-oklahoma-alabama-jersey]
08:35:47 [Q] INFO Processed [nineteen-winter-september-bravo]
08:35:47 [Q] INFO Processed [oscar-rugby-snake-glucose]
08:35:47 [Q] INFO Processed [thirteen-oxygen-sink-carpet]
08:35:47 [Q] INFO Processed [magnesium-video-xray-double]
08:35:47 [Q] INFO Processed [oranges-sweet-maine-monkey]
08:35:47 [Q] INFO Processed [tango-whiskey-connecticut-ack]
08:35:47 [Q] INFO Processed [king-twelve-vegan-dakota]
08:35:47 [Q] INFO Processed [happy-wolfram-fanta-sodium]
08:35:48 [Q] INFO Processed [georgia-wolfram-fix-washington]
08:35:48 [Q] INFO Processed [yellow-arizona-colorado-oregon]
08:35:48 [Q] INFO Processed [massachusetts-lion-sodium-batman]
08:35:48 [Q] INFO Processed [mexico-georgia-tennessee-berlin]
08:35:48 [Q] INFO Processed [two-xray-idaho-don]
08:35:49 [Q] INFO Processed [princess-oxygen-nuts-oscar]
08:35:49 [Q] INFO Processed [oscar-asparagus-sodium-california]
08:35:49 [Q] INFO Processed [fix-crazy-paris-december]
08:35:49 [Q] INFO Processed [fish-paris-red-jig]
08:35:49 [Q] INFO Processed [ten-johnny-seventeen-august]
08:35:49 [Q] INFO Processed [ink-maine-bacon-spaghetti]
08:35:49 [Q] INFO Processed [cup-alaska-spaghetti-kitten]
08:35:50 [Q] INFO Processed [golf-harry-oven-october]
08:35:50 [Q] INFO Processed [burger-winter-foxtrot-august]
08:35:50 [Q] INFO Processed [twelve-hydrogen-cardinal-music]
08:35:50 [Q] INFO Processed [bacon-queen-xray-burger]
08:35:50 [Q] INFO Processed [east-floor-four-colorado]
08:35:50 [Q] INFO Processed [idaho-fifteen-fourteen-skylark]
08:35:51 [Q] INFO Processed [idaho-cola-september-delaware]
08:35:51 [Q] INFO Processed [comet-yellow-spaghetti-black]
08:35:51 [Q] INFO Processed [fourteen-carbon-sierra-saturn]
08:35:51 [Q] INFO Processed [venus-alabama-bulldog-butter]
08:35:51 [Q] INFO Processed [bacon-low-shade-mississippi]
08:35:51 [Q] INFO Processed [oregon-angel-grey-beer]
08:35:52 [Q] INFO Processed [nineteen-utah-lithium-emma]
08:35:52 [Q] INFO Processed [alpha-salami-enemy-whiskey]
08:35:52 [Q] INFO Processed [king-oven-fourteen-sixteen]
08:35:52 [Q] INFO Processed [pizza-lion-jig-thirteen]
08:35:52 [Q] INFO Processed [uranus-single-lamp-three]
08:35:53 [Q] INFO Processed [violet-maryland-potato-texas]
08:35:53 [Q] INFO Processed [sad-one-colorado-georgia]
08:35:53 [Q] INFO Processed [music-july-georgia-golf]
08:35:53 [Q] INFO Processed [sweet-four-thirteen-gee]
08:35:53 [Q] INFO Processed [green-snake-network-arizona]
08:35:53 [Q] INFO Processed [fanta-jupiter-maine-sixteen]
08:35:53 [Q] INFO Processed [oklahoma-may-lactose-twenty]
08:35:53 [Q] INFO Processed [wisconsin-orange-nineteen-bluebird]
08:35:53 [Q] INFO Processed [yankee-floor-item-floor]
08:35:53 [Q] INFO Processed [mexico-maryland-carolina-king]
08:35:54 [Q] INFO Processed [sixteen-washington-apart-juliet]
08:35:54 [Q] INFO Processed [saturn-uncle-michigan-white]
08:35:54 [Q] INFO Processed [blue-failed-mike-ohio]
08:35:54 [Q] INFO Processed [uniform-vermont-pizza-nebraska]
08:35:54 [Q] INFO Processed [high-magazine-red-spring]
08:35:55 [Q] INFO Processed [carpet-thirteen-sixteen-nuts]
08:35:55 [Q] INFO Processed [texas-magnesium-beryllium-magnesium]
08:35:55 [Q] INFO Processed [winner-ceiling-queen-april]
08:35:55 [Q] INFO Processed [california-oxygen-chicken-nuts]
08:35:55 [Q] INFO Processed [fix-connecticut-triple-beryllium]
08:35:55 [Q] INFO Processed [jig-ink-lamp-oven]
08:36:44 [Q] INFO Enqueued 33366
08:36:44 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:36:44 [Q] INFO Process-1:64 processing [iowa-comet-cold-virginia]
08:36:44 [Q] INFO Processed [iowa-comet-cold-virginia]
08:39:44 [Q] INFO Enqueued 33367
08:39:44 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:39:44 [Q] INFO Enqueued 33368
08:39:44 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:39:44 [Q] INFO Enqueued 33369
08:39:44 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:39:44 [Q] INFO Enqueued 33370
08:39:44 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:39:44 [Q] INFO Enqueued 33371
08:39:44 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:39:45 [Q] INFO Process-1:66 processing [nevada-kansas-magnesium-quiet]
08:39:45 [Q] INFO Process-1:67 processing [island-social-island-yellow]
08:39:45 [Q] INFO Process-1:65 processing [delaware-gee-india-hydrogen]
08:39:45 [Q] INFO Process-1:64 processing [delta-iowa-bluebird-massachusetts]
08:39:45 [Q] INFO Process-1:64 processing [mockingbird-fifteen-alpha-music]
08:39:45 [Q] INFO Processed [delta-iowa-bluebird-massachusetts]
08:39:45 [Q] INFO Processed [nevada-kansas-magnesium-quiet]
08:39:46 [Q] INFO Processed [island-social-island-yellow]
08:39:46 [Q] INFO Processed [delaware-gee-india-hydrogen]
08:39:46 [Q] INFO Processed [mockingbird-fifteen-alpha-music]
08:41:45 [Q] INFO Enqueued 33372
08:41:45 [Q] INFO Process-1 created a task from schedule [InvenTree.tasks.heartbeat]
08:41:45 [Q] INFO Process-1:66 processing [red-arkansas-burger-oregon]
08:41:45 [Q] INFO Processed [red-arkansas-burger-oregon]
```

### PR DESCRIPTION
**Pull request by lippoliv**
#5606 improved amount of events already, but did not finally fix #5594. @SchrodingersGat came up with the point, that all events are processed by all listening plugins. I have two of those, and even if their `process_event` is guarded

```python
    def process_event(self, event, *args, **kwargs):
        if event != "salesordershipment.completed":
            return
```

the background worker will roughly spend 1 second per event per plugin. Stock movements are "event-heavy" and most of the events at least my plugins are not interested in, so here we loose a lot of time, probably.

# Idea of this change
Introduce `EventMixin.wants_process_event`, a simple function that returns `True` or `False`. `register_event` will call that function before `offload_task` and just continue if the plugin returns `True`.

The default implementation of `EventMixin.wants_process_event` will always return true to guarantee backwards compatiblity.

I also implemented tests, but I'm not sure if I read the output correctly on my machine. The code change is not big, so I hope my brain processed it correctly :)

# Side changes
I reduced cognitive complexity in `register_event` a little, by inverting to `if`s and call `continue` to guard the following code, instead of increase nesting.

# Backporting
I would love to see this being backported to 0.12.x

### CODE REVIEW

**Comment by lippoliv** [comment]
I patched it to my "production" environment and it reduces workload significantly. To visualize it, I added an extra line starting with the `----`es.

The amount of reduced background work multiplies with every plugin using `EventMixin`, so I think this is a serious performance optimization.

```
07:11:53 [Q] INFO Process-1:4 processing [blossom-comet-wisconsin-wisconsin]
07:11:53 [Q] INFO Process-1:1 processing [thirteen-mike-video-ink]
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stockitem.assignedtocustomer' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-provision-sales-order
07:11:53 [Q] INFO Process-1:4 processing [uranus-july-river-rugby]
-------- Ignoring triggering event 'stockitem.assignedtocustomer' for mj-auto-provision-sales-order
07:11:53 [Q] INFO Process-1:1 processing [eighteen-december-item-mockingbird]
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'order_salesorderlineitem.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'order_salesorderlineitem.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'order_salesorder.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'order_salesorder.saved' for mj-auto-provision-sales-order
07:11:53 [Q] INFO Processed [edward-alanine-two-oranges]
07:11:53 [Q] INFO Processed [foxtrot-tennis-william-alpha]
07:11:54 [Q] INFO Processed [lemon-six-potato-ten]
07:11:54 [Q] INFO Process-1:3 processing [pluto-mirror-monkey-don]
07:11:54 [Q] INFO Process-1:4 processing [artist-neptune-charlie-victor]
07:11:54 [Q] INFO Process-1:1 processing [texas-hot-diet-montana]
07:11:54 [Q] INFO Process-1:2 processing [ohio-charlie-maryland-item]
-------- Ignoring triggering event 'order_salesorderallocation.saved' for mj-auto-finish-sales-order
Updating part pricing for 17 | Aufkleber Cathrin Müller - 
-------- Ignoring triggering event 'order_salesorderallocation.saved' for mj-auto-provision-sales-order
07:11:54 [Q] INFO Process-1:3 processing [august-colorado-twelve-nine]
-------- Ignoring triggering event 'stock_stockitem.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitem.created' for mj-auto-provision-sales-order
07:11:54 [Q] INFO Process-1:4 processing [georgia-purple-apart-wisconsin]
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order
07:11:54 [Q] INFO Process-1:3 processing [pennsylvania-purple-sierra-tennessee]
07:11:54 [Q] INFO Process-1:4 processing [saturn-nitrogen-xray-violet]
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order
07:11:54 [Q] INFO Process-1:3 processing [solar-carpet-jig-leopard]
07:11:54 [Q] INFO Processed [fourteen-robert-missouri-music]
07:11:54 [Q] INFO Process-1:1 processing [tango-south-stairway-kitten]
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-provision-sales-order
07:11:55 [Q] INFO Processed [crazy-low-mountain-lamp]
07:11:55 [Q] INFO Processed [romeo-comet-moon-johnny]
07:11:55 [Q] INFO Process-1:1 processing [victor-snake-kitten-quebec]
07:11:55 [Q] INFO Process-1:3 processing [butter-king-kansas-pennsylvania]
07:11:55 [Q] INFO Process-1:4 processing [fillet-twelve-freddie-oscar]
07:11:55 [Q] INFO Process-1:2 processing [item-undress-aspen-minnesota]
07:11:55 [Q] INFO Processed [georgia-enemy-ceiling-autumn]
-------- Ignoring triggering event 'stockitem.assignedtocustomer' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stockitem.assignedtocustomer' for mj-auto-provision-sales-order
07:11:56 [Q] INFO Process-1:2 processing [stream-eighteen-blossom-ten]
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order
07:11:56 [Q] INFO Process-1:1 processing [charlie-sink-lake-carpet]
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-provision-sales-order
07:11:56 [Q] INFO Process-1:4 processing [helium-vegan-johnny-iowa]
-------- Ignoring triggering event 'order_salesorder.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'order_salesorder.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'order_salesorderlineitem.saved' for mj-auto-finish-sales-order
07:11:56 [Q] INFO Process-1:1 processing [pizza-east-video-ink]
-------- Ignoring triggering event 'order_salesorderlineitem.saved' for mj-auto-provision-sales-order
07:11:56 [Q] INFO Process-1:2 processing [robin-network-july-twelve]
-------- Ignoring triggering event 'order_salesorderallocation.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'order_salesorderallocation.saved' for mj-auto-provision-sales-order
07:11:56 [Q] INFO Process-1:4 processing [saturn-green-magazine-lake]
-------- Ignoring triggering event 'stock_stockitem.created' for mj-auto-finish-sales-order
Updating part pricing for 10 | Postkarte Römer 8, 28 | 1 - 
-------- Ignoring triggering event 'stock_stockitem.created' for mj-auto-provision-sales-order
07:11:56 [Q] INFO Processed [orange-wisconsin-gee-winter]
07:11:57 [Q] INFO Processed [emma-blossom-lithium-diet]
07:11:58 [Q] INFO Processed [johnny-equal-zebra-green]
07:11:58 [Q] INFO Process-1:1 processing [nine-white-princess-jersey]
07:11:58 [Q] INFO Process-1:2 processing [monkey-florida-twelve-sink]
07:11:58 [Q] INFO Process-1:3 processing [victor-edward-east-four]
07:11:58 [Q] INFO Process-1:4 processing [lake-golf-saturn-dakota]
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-provision-sales-order
07:11:58 [Q] INFO Process-1:3 processing [glucose-single-hawaii-ohio]
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order
07:11:59 [Q] INFO Process-1:2 processing [fish-kitten-batman-minnesota]
07:11:59 [Q] INFO Process-1:1 processing [cup-florida-mars-lake]
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-finish-sales-order
07:11:59 [Q] INFO Process-1:3 processing [saturn-item-eleven-wyoming]
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitem.saved' for mj-auto-provision-sales-order
07:11:59 [Q] INFO Process-1:1 processing [finch-aspen-idaho-salami]
07:11:59 [Q] INFO Process-1:2 processing [vegan-hot-fanta-football]
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.saved' for mj-auto-provision-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-finish-sales-order
-------- Ignoring triggering event 'stock_stockitemtracking.created' for mj-auto-provision-sales-order

```

**Review by matmair — APPROVED** [review]
I like the approach in general, nice work. I think this optimisation step should be documented in the plugin docs: https://github.com/inventree/InvenTree/blob/master/docs/docs/extend/plugins/event.md

Additionally (if you have the bandwidth) I think It would make sense to register the events a plugin is interested in only on startup in a dict or [djangos build-in cache](https://docs.djangoproject.com/en/4.2/topics/cache/#basic-usage). This would serve two interests:
1. Reading from the cache is very fast (more performance)
2. It would (if documented) clearly indicate where logic should be: in process_event. That way the intended performance gains are more likely as the current pattern could lead a developer to implement a db-lookup in the registering event and then again in the processing event

**Comment by matmair** [comment]
I do not think this qualifies for a backport. We only do that for bug fixes and this is not a bug fix.

File: InvenTree/plugin/base/event/events.py [hunk 1]
```diff
@@ -65,27 +65,32 @@ def register_event(event, *args, **kwargs):
...
+                if not plugin.is_active():
+                    continue
 
-                    if plugin.is_active():
-                        # Only allow event registering for 'active' plugins
+                if not plugin.wants_process_event(event):
+                    continue
 
-                        logger.debug(f"Registering callback for plugin '{slug}'")
+                # Only allow event registering for 'active' plugins
```
[the remark this answers is not in the document]
Comment (lippoliv) [reply InvenTree/plugin/base/event/events.py:77]: Are you OK if I remove that comment? I strongly beliefe "Comments are lies waiting to happen". The code is kind of self-explaining, so I would go without the comment.

If you like to have it, sure I can move it.

[the remark this answers is not in the document]
Comment (matmair) [reply InvenTree/plugin/base/event/events.py:77]: not really, comments serve to help new contributors to understand the code. Not updating comments is on the coder that changes the code

**Comment by lippoliv** [comment]
> I think this optimisation step should be documented in the plugin docs

👍  I'll try

> Additionally (if you have the bandwidth) I think It would make sense to register the events a plugin is interested in only on startup in a dict or [djangos build-in cache](https://docs.djangoproject.com/en/4.2/topics/cache/#basic-usage).

👍 Sounds interesting, I'll give it a try.

> I do not think this qualifies for a backport. We only do that for bug fixes and this is not a bug fix.

First: OK for me, I found a way to deploy it to my NAS (mount in modified files). 

🤔 From my experience, performance issues can become a blocker faster than you think. That's why I would (and did with #5594) classify it as a bug. Sure, "pay" (more hardware) always hides performance issues, but not more.

So I would go, if possible, with the backport as long as it doesn't postpone 0.12.8.
If 0.13.0 is available soon, it's probably irrelevant anyways.

**Comment by matmair** [comment]
> From my experience, performance issues can become a blocker faster than you think

I manage or support around 50 instances, this has never been a problem. Breaking the policy will end in everyone requesting backports, making the stable release channel less stable ie. defeating the purpose.

**Comment by wolflu05** [comment]
General question: what do you think of decorators for registering event handlers? I use them a lot in other languages. I now read, Django has built in signals that use decorators, why not use them? @matmair

**Comment by lippoliv** [comment]
> Breaking the policy will end in everyone requesting backports, making the stable release channel less stable ie. defeating the purpose.

That's indeed a good argument.

**Comment by SchrodingersGat** [comment]
@lippoliv nice implementation here, you certainly understood the main thrust of my suggestion in the other thread.

### Caching

@matmair has suggested some caching which initially I agreed with. However, plugins may need to determine dynamically if they will support a particular event. So I think that caching will not work. The actual "expense" of this check is pretty minimal, I do not think it will be a significant performance hit

### Documentation

Definitely this will need some notes in the plugin documentation

### Backwards Compatibility

I like your approach to backwards compatibility - I would also like to encourage plugin developers to implement the new check function. Can you add into the check function a `DeprecationWarning` - indicating that this function will be required in a future release.

**Comment by lippoliv** [comment]
@matmair @SchrodingersGat I guess whether to backport or not you can clarify "internally" after this PR is done :)

> The actual "expense" of this check is pretty minimal, I do not think it will be a significant performance hit

Yes, it's not async (which seems to be the problem with `process_event`) so it's one function-call + whatever the DEV is doind (it should be small code). I currently have all my problems gone.

> However, plugins may need to determine dynamically if they will support a particular event.

I think this can become a "conflict": The code whether the plugin wants a event or not, is executed synchronously. It should be performant :)

However all events are sent to `process_event`, at that moment it's the worker which executes it asynchronously. So here we can be a little more expensive when it comes to code.

I would say `wants_process_event` is kind of static, just reducing "all events" down to "this 5" and then `process_event` maybe still skips some events. That should be a fair tradeoff.

**Conclusion**: I think caching (or decorators mentioned by @wolflu05) would be fine.

> Can you add into the check function a DeprecationWarning - indicating that this function will be required in a future release.

I don't understand. Nothing is deprecated?

**Comment by SchrodingersGat** [comment]
> Conclusion: I think caching (or decorators mentioned by @wolflu05) would be fine.

I think what I am saying has some subtle differences. A given plugin may not want to *always* respond to a certain event - perhaps depending on a plugin setting.

However, for simplicity, we could certainly cache the information of "which plugins want which event" and let any subsequent event-handling decisions to be managed in the background worker thread

**Comment by SchrodingersGat** [comment]
> I don't understand. Nothing is deprecated?

Not yet - but this message will encourage plugin developers to implement the check function.

**Comment by wolflu05** [comment]
If we're using decorators, they can also be invoked as normal functions. That way we can also dynamically register events at runtime and cache that, just by invoking the decorator with the handler function and event strings that it wants to process somewhere in the plugin.

**Comment by SchrodingersGat** [comment]
@wolflu05 moving to decorators would be nice and "pythonic" - but not compatible with existing plugin implementations

**Comment by wolflu05** [comment]
Why not, we could deprecate the process event function and if a plugin still defined it, we can just register it for all events on plugin startup for a few releases until the plugins get adjusted.

**Comment by SchrodingersGat** [comment]
@wolflu05 fair call - can you provide an example implementation of what you are imagining this might look like?

**Comment by wolflu05** [comment]
Maybe something like this, but not sure how we should name the unhandle or disconnected event function.

```py
from plugin import EventMixin, InvenTreePlugin, handle_event
from stock.models import StockLocation

class MyPlugin(EventMixin, InvenTreePlugin):
    def __init__(self, *args, **kwars):
        super().__init__(*args, **kwars)
        if whatever_condition:
            handle_event(self.log_part_category_save, model_PartCategory, event="post_save")
        else:
            dissconnect_event(self.log_part_category_save, model_PartCategory, event="post_save")

    @handle_event(model=StockLocation, event="post_save"):
    def log_created_stock_locations(self, instance, **kwargs):
        print(f"Saved location {instance}")

    def log_part_category_save(self, instance, **kwargs):
        print(f"Saved category {instance}")
```

**Comment by SchrodingersGat** [comment]
@wolflu05 I think there are some good ideas there, but might overly complicate what @lippoliv is trying to achieve here.

My vote is that we merge this simple update, which provides a path to reducing CPU load without causing any compatibility issues.

Any further updates (such as function decorators) should be pushed to a different future PR

**Comment by matmair** [comment]
I like @wolflu05 suggestion but that should propably be a separate PR (where we can also implement caching).

I am inline with @SchrodingersGat and would suggest that we wait for the docs for the new function to be added and merge, then do the more advanced things.

File: InvenTree/plugin/base/event/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lippoliv) [reply InvenTree/plugin/base/event/events.py:77]: I didn't ignore it willingly. Fixed it now.

**Comment by lippoliv** [comment]
@matmair @SchrodingersGat I merged in master and added documentation, happy to see some feedback :)

**Comment by SchrodingersGat** [comment]
@lippoliv nice work, thanks for the contribution. When implemented correctly this should really reduce load on the worker

**Pull request merged**